docs/code/SchemaComprehension.md

```diff
@@ -0,0 +1,210 @@
+| `U1`             | `byte`      | `byte?`                        |
+| `U2`             | `ushort`    | `ushort?`                      |
+| `U4`             | `uint`      | `uint?`                        |
+| `U8`             | `ulong`     | `ulong?`                       |
+| `UG`             | `UInt128`   |                                |
+| `R4`             | `float`     | `float?`                       |
+| `R8`             | `double`    | `double?`                      |
+| `TX`             | `DvText`, `string` |                         |
+| `BL`             | `DvBool`           | `bool`, `bool?`         |
+| `TS`             | `DvTimeSpan`       |                         |
+| `DT`             | `DvDateTime`       |                         |
+| `DZ`             | `DvDateTimeZone`   |                         |
+| Variable-size vector | `VBuffer<T>`   | `T[]`, and the vector is always dense |
+| Fixed-size vector    | `VBuffer<T>` with `[VectorType(N)]` | `T[]` with `VectorType(N)`, and the vector is always dense |
+| Key type             | `uint` with `[KeyType]`             |                                                            |
+
+### Additional attributes to affect type mapping
+
+There are two more attributes that can affect the way ML.NET conducts schema comprehension:
+* `[ColumnName]` lets you choose a different name for the `IDataView` column. By default it is the same as field name.
+  * This is a way to create or read back an `IDataView` column with a name containing 'invalid' characters (like whitespace).
+* `[NoColumn]` is an attribute that denotes that the below field should not be mapped to a column.
+
+### Using SchemaDefinition for run-time type mapping hints
+
+As you can see from the table and notes above, certain `IDataView` types can only be denoted with an additional field attribute. If the type parameters are not known at compile time (like the size of the fixed-size vector), this is tricky. 
+
+You can use a `SchemaDefinition` object to re-map a type to an `IDataView` schema programmatically. It gives you the same powers as the attributes, but at runtime.
+Please see the below example.
+```(csharp)
+// Vector size is only known at runtime.
+int numberOfFeatures = 4;
+
+// Create the default schema definition.
+var schemaDef = SchemaDefinition.Create(typeof(IrisVectorData));
+
+// Specify the right vector size.
+schemaDef["Features"].ColumnType = new VectorType(NumberType.R4, numberOfFeatures);
+
+// Create a data view.
+var dataView = env.CreateDataView<IrisVectorData>(arr, schemaDef);
+
+// Create a prediction engine. You can add custom input and output schema definitions there.
+var predictionEngine = env.CreatePredictionEngine<IrisData, IrisVectorData>(dv, outputSchemaDefinition: schemaDef);
+```
+
+In addition to the above, you can use `SchemaDefinition` to add per-column metadata, or even a 'value generator' (so that the column value is not read from the field, but computed using a delegate).
```

Comment:
It might be interesting to have a code snippet example for this scenario?

Reply:
In fact, I think the 'generator' bit didn't make it into ML.NET

---
In reply to: [204536819](https://github.com/dotnet/machinelearning/pull/572#discussion_r204536819) [](ancestors = 204536819)